drivers/ieee802154/ieee802154_nrf5.c

```diff
@@ -653,6 +659,28 @@ static void nrf5_irq_config(const struct device *dev)
 #endif
 }
 
+#if defined(CONFIG_OPENTHREAD_TIME_SYNC)
+void nrf_802154_tx_started(const uint8_t *aFrame)
```

Comment:
I'm afraid this won't work for nRF53 platform. AFAIK `nrf_802154_tx_started` will be eventually removed for all platforms. Notice `nrf5_tx_started` already exists.

Also there will be conflicts when CSL or Enh-Ack Link Metrics add Information Elements to the frame as well.

Ideally, the Time IE filling should be handled by the radio driver itself, providing a configuration API to be used in this shim layer.

Reply:
Does `nrf5_tx_started` actually serve similar purpose as `nrf_802154_tx_started`?
By conflicts with CSL you mean that they would arise if it's implementation would also use `nrf_802154_tx_started`?
Is handling of Information Elements filling implemented in radio driver now, or will it be in the future?

Reply:
@czeslawmakarski can you please add some more light in here?

Reply:
There is a 802.15.4 Radio Driver callout `nrf_802154_tx_started`, which is being called on the actual start of the frame in the air, and, there is a Zephyr Shim function `nrf5_tx_started` which is being called when the Radio Driver did schedule the transmission of the frame. These two moments are not necessarily close to each other in time, since, for example, during the CSMA-CA procedure there can be numerous backoffs before actual frame transmission (not mentioning the Delayed TX aka TX_TIME and other use cases).

The `nrf5_tx_started` is being called on the AppCore. The `nrf_802154_tx_started` is being called on the same core as the whole Radio Driver runs (in case of nRF53 it's NetCore), and it is not being serialized (serialization would take so much time that it would make no sense for that, since frame shall be already in the air at least partly). Also, modifying the payload of the frame in the shim in `nrf5_tx_started` would also not make any sense in the nRF53, since frames are being memcopied from AppCore to the NetCore during the schedule of transmission, and it is the NetCore which alters them by injecting proper IEs and Link Metrics and other dynamic data.

From what I can see in this PR, the goal was to inject some node own time into the IE element. The best way to achieve that would be to add some call to the internals of the Radio Driver (then we can have the best possible accuracy). If one does not care that much about the accuracy, it would be needed to inject that data before calling any of the functions which do schedule the transmission (`nrf_802154_transmit_raw` and others).

Tagging some experts for visibility: @e-rk, @jciupis

Reply:
I wholeheartedly agree with @czeslawmakarski on this matter. I'd like to add though that this solution might be prone to bugs on nRF52 series as well. Firstly, as follows from this function's documentation:
```
* @note This function should be very short to prevent dropping frames by the driver
```
Calling `nrf5_get_time` potentially makes it rather lengthy in terms of execution time. Secondly, modifying the frame's MAC header content this late will lead to problems with encryption. The driver does not take into account that the API user might keep modifying the frame's content after submitting transmission request. If the frame is to be encrypted, the frame modifications performed here most probably (would require a double-check) will not be transmitted at all.

Reply:
Is there an already implemented example of filling some other delay-sensitive data that might help me find a way to do it correctly?
What is the API for configuring IE filling in radio driver that could be used here?

Reply:
There is an implementation for writing CSL phase and period into information elements:
https://github.com/nrfconnect/sdk-nrfxlib/blob/48668a92e583fc7c8e5497bc8f051edf3fd65b9d/nrf_802154/driver/src/mac_features/nrf_802154_ie_writer.c#L72

One important thing to consider is, does your code for timesync require some interaction with OpenThread during the TX phase, or can it work totally independently? Since the code would have to execute on a very high interrupt priority and in case of nRF53 it would be executing on the netcore, you won't have an option to interact with OpenThread, or anything else that runs on the application core.